protobuf_serialization/stdlib_readers.nim

```diff
@@ -0,0 +1,186 @@
+#Included by reader.
+
+import sets
+
+import stew/shims/macros
+import faststreams
+
+import internal
+import types
+
+proc decodeNumber[T, E](
+  stream: InputStream,
+  next: var T,
+  encoding: typedesc[E]
+) =
+  var flattened: flatType(T)
+  when E is VarIntWrapped:
+    flattened = stream.decodeVarInt(type(flattened), encoding)
+  elif E is FixedWrapped:
+    stream.decodeFixed(flattened)
+  else:
+    {.fatal: "Trying to decode a number which isn't wrapped. This should never happen.".}
+  box(next, flattened)
+
+proc readValue*(reader: ProtobufReader, value: var auto)
+
+proc stdlibFromProtobuf[R](
+  stream: InputStream,
+  _: typedesc[R],
+  unusedFieldName: static string,
+  value: var string
+) =
+  value = newString(stream.totalUnconsumedBytes)
+  for c in 0 ..< value.len:
+    value[c] = char(stream.read())
+
+proc stdlibFromProtobuf[R](
+  stream: InputStream,
+  _: typedesc[R],
+  unusedFieldName: static string,
+  value: var cstring
+) =
+  var preValue = newString(stream.totalUnconsumedBytes)
+  for c in 0 ..< preValue.len:
+    preValue[c] = char(stream.read())
+  value = preValue
+
+proc stdlibFromProtobuf[R, T](
+  stream: InputStream,
+  ty: typedesc[R],
+  fieldName: static string,
+  seqInstance: var seq[T]
```

Comment:
I think arrays and sequences should have been treated with the semantics of the ProtoBuf `repeated` keyword.
That is, you keep going through the buffer and you keep finding elements with the same field number. The elements doesn't even have to appear consequently as far as ProtoBuf is concerned. Every time you find such an element, you append it to the array/sequence.

Reply:
Yep. This is being marked as required. My bad.

Reply:
This is... complicated. https://github.com/kayabaNerve/nim-protobuf-serialization/issues/29 covers the problem.